<div style="border:solid Chocolate 2px; padding: 40px">
 
<b> Даниил, привет!👋</b>  

Меня зовут Кирилл Васильев, я буду ревьюером твоего проекта. Я предлагаю общаться на «ты», но если привычнее на «вы», просто скажи об этом! 

Я буду оставлять комментарии в твоем проекте. Пожалуйста, не удаляй их, они могут быть полезны в случае повторной проверки.

Ты можешь реагировать на комментарии как просто написав их в Markdown-ячейках, так и выделив их цветом. Например, <font color='blue'>синим</font>. Второй способ, на мой взгляд, удобнее.

Свои комментарии я буду обозначать <font color='green'>зеленым</font>, <font color='gold'>желтым</font> и <font color='red'>красным</font> цветами, например:
 
<br/>

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> В случае, если решение на отдельном шаге является полностью правильным. Здесь же я могу давать советы и предложения.
</div>

<br/>

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> В случае, когда решение на отдельном шаге станет еще лучше, если внести небольшие коррективы.
</div>


<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку🤔:</b>
 В случае, когда решение на отдельном шаге требует существенной переработки и внесения правок. Напоминаю, что проект не может быть принят с первого раза, если ревью содержит комментарии, рекомендующие доработать шаги.
</div>
    
Увидев у тебя неточность, в первый раз я лишь укажу на ее наличие и дам тебе возможность самому найти и исправить ее. На реальной работе твой руководитель будет поступать также. Но если ты пока не справишься с такой задачей - при следующей проверке я дам более точную подсказку! 

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка данных</a></span></li><li><span><a href="#Исследование-задачи" data-toc-modified-id="Исследование-задачи-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Исследование задачи</a></span></li><li><span><a href="#Борьба-с-дисбалансом" data-toc-modified-id="Борьба-с-дисбалансом-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Борьба с дисбалансом</a></span></li><li><span><a href="#Тестирование-модели" data-toc-modified-id="Тестирование-модели-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Тестирование модели</a></span></li><li><span><a href="#Чек-лист-готовности-проекта" data-toc-modified-id="Чек-лист-готовности-проекта-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Чек-лист готовности проекта</a></span></li></ul></div>

# Проект по определению возможного оттока клиентов

Из «Бета-Банка» стали уходить клиенты. 

Нужно спрогнозировать, уйдёт клиент из банка в ближайшее время или нет. Нам предоставлены исторические данные о поведении клиентов и расторжении договоров с банком. 

Необходимо построить модель с предельно большим значением *F1*-меры. Минимально необходимое значение метрики равно 0.59.

Необходимо дополнительно измерить *AUC-ROC* и сравнить её значение с *F1*-мерой.

Источник данных: [https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling](https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling)

Признаки данных:
```

RowNumber — индекс строки в данных
CustomerId — уникальный идентификатор клиента
Surname — фамилия
CreditScore — кредитный рейтинг
Geography — страна проживания
Gender — пол
Age — возраст
Tenure — сколько лет человек является клиентом банка
Balance — баланс на счёте
NumOfProducts — количество продуктов банка, используемых клиентом
HasCrCard — наличие кредитной карты
IsActiveMember — активность клиента
EstimatedSalary — предполагаемая зарплата

```

Целевой признак

```
Exited — факт ухода клиента

```

## Подготовка данных

Установим последнюю версию scikit-learn:

In [1]:
pip install -U scikit-learn

     |████████████████████████████████| 9.6 MB 1.3 MB/s eta 0:00:01        | 2.6 MB 1.0 MB/s eta 0:00:07     |█████████████▊                  | 4.1 MB 1.0 MB/s eta 0:00:06     |████████████████████████████▊   | 8.6 MB 1.3 MB/s eta 0:00:01     |████████████████████████████████| 9.6 MB 1.3 MB/s eta 0:00:01
     |████████████████████████████████| 296 kB 2.9 MB/s eta 0:00:010:01     |████████████████████████████████| 297 kB 2.9 MB/s 
  Attempting uninstall: joblib
    Found existing installation: joblib 1.1.0
    Uninstalling joblib-1.1.0:
      Successfully uninstalled joblib-1.1.0
  Attempting uninstall: scikit-learn
    Found existing installation: scikit-learn 0.24.1
    Uninstalling scikit-learn-0.24.1:
      Successfully uninstalled scikit-learn-0.24.1
Note: you may need to restart the kernel to use updated packages.


Установим Imbalanced-learn:

In [2]:
!pip install imblearn

     |████████████████████████████████| 225 kB 1.1 MB/s eta 0:00:01     |█████████████                   | 92 kB 997 kB/s eta 0:00:01     |████████████████████████████████| 226 kB 1.1 MB/s 


Импортируем необходимые библиотеки:

In [3]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.metrics import f1_score, roc_auc_score, recall_score 
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.utils import shuffle
from imblearn.over_sampling import SMOTE
from sklearn.dummy import DummyClassifier

<div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 
    
Если загружаем несколько функций из одного пакета, то удобнее записать это через запятую.
    
`from sklearn.metrics import f1_score, roc_auc_score`
    
А если через запятую будет больше трех позиций, в таком случае здесь было бы идеально привести оформление в соответствие с <a href="https://pythonworld.ru/osnovy/pep-8-rukovodstvo-po-napisaniyu-koda-na-python.html">pep8</a> - не оставлять длинных строк. 

Такое оформление предпочтительнее:    
    
`from sklearn.metrics import (
    accuracy_score, 
    confusion_matrix, 
    f1_score, 
    mean_squared_error, 
    roc_auc_score
)`
    
        
</div>


<div class="alert alert-info">
✍️ Учел, поправил
    

        
        


<div class="alert alert-success">
<h2> Комментарий ревьюера #2 <a class="tocSkip"> </h2>

👍 Cнова привет. Отлично. Спасибо за оформление :)
</div>

Загрузим рабочий датасет:

In [4]:
data = pd.read_csv('/datasets/Churn.csv')

Изучим:

In [5]:
data.head(5)

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0     2.0       0.00              1          1               1   
1     1.0   83807.86              1          0               1   
2     8.0  159660.80              3          1               0   
3     1.0       0.00              2          0               0   
4     2.0  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           9091 non-null   float64
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


В данных есть информация о 10000 клиентов, сразу видно, что есть пропуски в графе `Tenure`. Так же в рамках нашего исследования нам не понадобятся некоторые колонки - индекс строки, уникальный идентификатор клиента и его фамилия. Эти колонки можно удалить. А вот с кредитным рейтингом, страной проживания, полом, возрастом, тем сколько клиент работает с банком, его баланс на счете, количество продуктов, наличие кредитной карты, его активность и зарплата это то, с чем мы будем работать.

Проведем проверку на дубликаты:

In [7]:
data.duplicated().sum()

0

Дубликатов нет.

Как отметили раньше в данных пропуски в столбце `Tenure`, удалим их из наших данных:

In [8]:
data = data.dropna().reset_index(drop=True)

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Согласен
</div> 


Теперь приведем названия столбцов в соответствии с принятым стилем:

In [9]:
data = data.rename(columns={'RowNumber':'row_number', 'CustomerId':'customer_id', 'CreditScore':'credit_score', 'NumOfProducts':'num_of_products', 'HasCrCard':'has_cr_card', 'IsActiveMember':'is_active_member', 'EstimatedSalary':'estimated_salary', })

In [10]:
data.columns= data.columns.str.lower()

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Чтобы не делать вручную, можно использовать регулярные выражения. Например, так:

`df.columns = [re.sub(r'(?<!^)(?=[A-Z])', '_', i).lower() for i in df.columns]`
    
Подробнее про регулярные выражения можно почитать здесь: https://habr.com/ru/post/349860/    
</div>


Теперь удалим колонки, которые не подойдут для дальнешего исследования:

In [11]:
data.drop(['row_number', 'customer_id', 'surname'], axis=1, inplace=True)

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Всё верно
</div>

    
    

In [12]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9091 entries, 0 to 9090
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   credit_score      9091 non-null   int64  
 1   geography         9091 non-null   object 
 2   gender            9091 non-null   object 
 3   age               9091 non-null   int64  
 4   tenure            9091 non-null   float64
 5   balance           9091 non-null   float64
 6   num_of_products   9091 non-null   int64  
 7   has_cr_card       9091 non-null   int64  
 8   is_active_member  9091 non-null   int64  
 9   estimated_salary  9091 non-null   float64
 10  exited            9091 non-null   int64  
dtypes: float64(3), int64(6), object(2)
memory usage: 781.4+ KB


Теперь, когда мы избавились от пропусков, подтвердили отсутствие дубликатов и убали ненужные столбцы и стилевые нарушения можно начать наше исследование.

## Исследование задачи

Для того, чтобы обучить модель и проверить ее работу понадобится разбить данные на три выборки: учебную, валидационную и тестовую. Но перед этим разобьем выборку на признаки и целевые признаки. В целевые признаки `target` возьмем столбец с информацией о том ушел клиент или нет, в обычные признаки `features` пойдут все остальные столбцы:

In [13]:
features = data.drop(['exited'], axis=1)
target = data['exited']

Для разбития данных на выборки используем `train_test_split`. Сначала разобьем на тренировочную выборку `train` и выборку для проверки `check` в соотношении 60% и 40%, потом выборку для проверки `check` разобьем пополам, чтобы получилось по 20% от всей выборки на валидационную `valid` и тестовую `test` выборки:

In [14]:
features_train, features_check, target_train, target_check, = train_test_split(
    features, target, test_size=0.40, random_state=12345)

In [15]:
features_valid, features_test, target_valid, target_test = train_test_split(
    features_check, target_check, test_size=0.50, random_state=12345)

Проверим, чтобы количественно выборки соответствовали необходимым параметрам:

In [16]:
print(features_train.shape)
print(target_train.shape)
print(features_valid.shape)
print(target_valid.shape)
print(features_test.shape)
print(target_test.shape)

(5454, 10)
(5454,)
(1818, 10)
(1818,)
(1819, 10)
(1819,)


<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 С разбиением на выборки согласен. 
</div>


Выборки подготовленны.

Следующим шагом будет преобразование категориальных признаков в численные, для этого используем возможности `OneHotEncoder`. Для начала определим, с какими столбцами будем работать:

In [17]:
categorical = ['geography', 'gender']

Теперь обучим функцию на нашей тренировочной выборке.

In [18]:
ohe = OneHotEncoder(handle_unknown='ignore',sparse_output=False, drop='first')

ohe.fit(features_train[categorical])

OneHotEncoder(drop='first', handle_unknown='ignore', sparse_output=False)

И перекодируем каждую нашу выборку с помощью `OneHotEncoder`:

Тренировочную:

In [19]:
features_train_encoded = features_train.copy()

features_train_encoded[ohe.get_feature_names_out()] = ohe.transform(features_train[categorical])

features_train_encoded = features_train_encoded.drop(categorical, axis=1)

Валидационную:

In [20]:
features_valid_encoded = features_valid.copy()

features_valid_encoded[ohe.get_feature_names_out()] = ohe.transform(features_valid[categorical])

features_valid_encoded = features_valid_encoded.drop(categorical, axis=1)

И тестовую:

In [21]:
features_test_encoded = features_test.copy()

features_test_encoded[ohe.get_feature_names_out()] = ohe.transform(features_test[categorical])

features_test_encoded = features_test_encoded.drop(categorical, axis=1)

Проверим как теперь выглядят столбцы на примере тренировочной выборки:

In [22]:
features_train_encoded.head(3)

credit_score  age  tenure    balance  num_of_products  has_cr_card  \
8483           727   28     2.0  110997.76                1            1   
3431           537   26     7.0  106397.75                1            0   
6770           610   40     9.0       0.00                1            1   

      is_active_member  estimated_salary  geography_Germany  geography_Spain  \
8483                 0         101433.76                0.0              0.0   
3431                 0         103563.23                0.0              0.0   
6770                 1         149602.54                0.0              0.0   

      gender_Male  
8483          0.0  
3431          1.0  
6770          1.0

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Супер. Все абсолютно корректно. Ничего не забыли.
</div>


Для дальнейшей работы масштабируем наши выборки. Сначала выберем численные признаки для масштабирования:

In [23]:
numeric = ['credit_score','age','tenure','balance','num_of_products','has_cr_card','is_active_member','estimated_salary']

Теперь создадим объект масштабирования и настроим его на обучающей выборке:

In [24]:
scaler = StandardScaler()
scaler.fit(features_train_encoded[numeric]) 
features_train_encoded[numeric] = scaler.transform(features_train_encoded[numeric])
features_valid_encoded[numeric] = scaler.transform(features_valid_encoded[numeric])
features_test_encoded[numeric] = scaler.transform(features_test_encoded[numeric])

Проверим получилось ли масштабировать по учебной выборке:

In [25]:
features_train_encoded.head(3)

credit_score       age    tenure   balance  num_of_products  \
8483      0.809075 -1.039327 -1.025995  0.554904        -0.908179   
3431     -1.152518 -1.227561  0.696524  0.480609        -0.908179   
6770     -0.398853  0.090079  1.385532 -1.237830        -0.908179   

      has_cr_card  is_active_member  estimated_salary  geography_Germany  \
8483     0.663468         -1.024127          0.019508                0.0   
3431    -1.507231         -1.024127          0.056167                0.0   
6770     0.663468          0.976442          0.848738                0.0   

      geography_Spain  gender_Male  
8483              0.0          0.0  
3431              0.0          1.0  
6770              0.0          1.0

Теперь начнем изучать какие модели лучше себя поведут:

Обучим и используем три модели:
- Модель дерево решений (Decision Tree Classifier)
- Модель случайного леса (Random Fores tClassifier)
- Модель логистической регрессии (Logistic Regression)

<div class="alert alert-block alert-danger">
<h2><s> Комментарий ревьюера <a class="tocSkip"></h2>
  
<b>На доработку🤔:</b>
 
Ниже используем линейные модели. Поэтому здесь необходимо сделать масштабирование. Напомню, делаем его после разбиения на выборки: обучаемся на тренировочной, а применяем ко всем трем, включая тестовую.

</div>


<div class="alert alert-info">
✍️ Масштабировал
    


Начнем с модели дерева решений `model_dtc`. Установим какой параметр максимальной глубины дерева `depth` (от 1 до 11) даст модель с  лучшим показателем *F1* при сравнении с ответами валидационной выборки.

In [26]:
best_result_dtc = 0
best_depth_dtc = 0

for depth in range(1, 11):
    model_dtc = DecisionTreeClassifier(random_state=12345, max_depth=depth)
    model_dtc.fit(features_train_encoded, target_train)
    predictions_valid_dtc = model_dtc.predict(features_valid_encoded)
    result_valid_dtc = f1_score(target_valid, predictions_valid_dtc)
    if result_valid_dtc > best_result_dtc:
        best_result_dtc = result_valid_dtc
        best_depth_dtc = depth
        
    
print(' F1-мера наилучшей модели дерева решений на валидационной выборке:', best_result_dtc, 
'\n При максимальной глубине дерева:', best_depth_dtc)

 F1-мера наилучшей модели дерева решений на валидационной выборке: 0.5764331210191083 
 При максимальной глубине дерева: 7


<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Отлично
</div>

Теперь очередь модели случайного леса `model_rfc`. Установим какой параметр количества деревьев `est` и какая глубина дерева `depth` даст модель с лучшим показателем *F1* при сравнении с ответами валидационной выборки.

In [27]:
best_result_rfc = 0
best_est_rfc = 0
best_depth_rfc = 0

for est in range(1, 16):
    for depth in range(10, 26):
        model_rfc = RandomForestClassifier(random_state=12345, n_estimators=est, max_depth=depth)
        model_rfc.fit(features_train_encoded, target_train)
        predictions_valid_rfc = model_rfc.predict(features_valid_encoded)
        result_valid_rfc = f1_score(target_valid, predictions_valid_rfc)
        if result_valid_rfc > best_result_rfc:
            best_result_rfc = result_valid_rfc
            best_est_rfc = est
            best_depth_rfc = depth
            

        
print(' F1-мера наилучшей модели случайного леса на валидационной выборке:', best_result_rfc,
      '\n При количестве деревьев:', best_est_rfc, 'При глубине деревьев', best_depth_rfc)

 F1-мера наилучшей модели случайного леса на валидационной выборке: 0.60625 
 При количестве деревьев: 13 При глубине деревьев 24


<div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

А здесь если мы подбираем сначала один гиперпараметр, а затем другой, то мы пропускаем несколько потенциально оптимальных комбинаций. Чуть дольше будет считаться, но зато лучший результат получим, если будем применять подбор гиперпараметров в двойном цикле:
    
`for est in range(10, 101, 10):
    for depth in range(1, 14, 2):
        model = RandomForestClassifier(random_state=12345, n_estimators=est, max_depth=depth)`
      
 
</div>



<div class="alert alert-info">
✍️ Поменял, но в итоге вышел такой же результат
    



<div class="alert alert-success">
<h2> Комментарий ревьюера #2 <a class="tocSkip"> </h2>

👍 Да, но теперь мы точно попробовали все комбинации и знаем, что не пропустили лучшую.
</div>

Теперь проверим модель логистической регрессии `model_lr`. Установим какой параметр итерации `itr` (от 100 до 1000, с шагом в 10) даст модель с лучшим показателем *F1*  при сравнении с ответами валидационной выборки.

In [28]:
best_result_lr = 0
best_itr_lr = 0

for itr in range(100, 501, 10):
    model_lr = LogisticRegression(random_state=12345, solver='liblinear', max_iter=itr)
    model_lr.fit(features_train_encoded, target_train)
    predictions_valid_lr = model_lr.predict(features_valid_encoded)
    result_valid_lr = f1_score(target_valid, predictions_valid_lr)
    if result_valid_lr > best_result_lr:
        best_result_lr = result_valid_lr
        best_itr_lr = itr

print(' F1-мера наилучшей модели логистической регрессии на валидационной выборке:', best_result_lr,
      '\n При количестве итераций:', best_itr_lr)

 F1-мера наилучшей модели логистической регрессии на валидационной выборке: 0.3033932135728543 
 При количестве итераций: 100


<div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

Сказывается отсутствие масштабирования.
    
По max_iter, думаю, улучшений особых не будет. Модель сама неплохо останавливается, когда достигает минимума.
       
</div>

<div class="alert alert-info">
✍️ Т.е. я так понимаю, что для логической регрессии нет особого смысла настраивать гиперпараметры?
    



<div class="alert alert-success">
<h2> Комментарий ревьюера #2 <a class="tocSkip"> </h2>

👍 Сложно ответить однозначно. В этой задаче смысл будет слабый для регрессии. Можно попробовать параметр «C».
</div>

Мы разбили наши данные на три выборки, необходимые для работы, а после поменяли категориальные признаки в численные. Затем после исследования трех различных моделей мы установиил, что лучше всего себя показала модель случайного леса с количеством деревьев 39 и глубиной 24. Так как эта модель лучшая, то после того, как уберем дисбаланс в учебной выборке, то мы снова ее обучим и проверим ее качество.

## Борьба с дисбалансом

Проверим есть ли дисбаланс классов:

In [29]:
np.unique(target_train, return_counts = True)

(array([0, 1]), array([4328, 1126]))

Дисбаланс присутствует, отрицательных значений больше в 4 раза. Для того чтобы убрать дисбаланс в данных воспользуемся двумя методами. Сначала создадим функцию, чтобы увеличить число положительных ответов, а затем создадим функцию, чтобы сократить количество отрицательных. В каждой из них перебором выберем необходимый параметр - насколько увеличить и насколько сократить. 

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Дисбаланс исследовали. Да, он достаточно сильный, чтобы негативно повлиять на качество моделей.
</div>



~Сначала создадим функцию для увеличения выборки:~

In [30]:
#def upsample(features, target, repeat):
#    features_zeros = features[target == 0]
#    features_ones = features[target == 1]
#    target_zeros = target[target == 0]
#    target_ones = target[target == 1]

#    features_upsampled = pd.concat([features_zeros] + [features_ones] * repeat)
#    target_upsampled = pd.concat([target_zeros] + [target_ones] * repeat)
    
#    features_upsampled, target_upsampled = shuffle(
#        features_upsampled, target_upsampled, random_state=12345)
    
#    return features_upsampled, target_upsampled

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Отлично, что используешь рукописную функцию, но здесь можно все оптимизировать, используя <a href="https://imbalanced-learn.org/stable/references/generated/imblearn.over_sampling.SMOTE.html">SMOTE</a> для увеличения выборки и <a href="https://imbalanced-learn.org/stable/references/generated/imblearn.under_sampling.RandomUnderSampler.html">RandomUnderSampler</a>  для уменьшения. 
    
Апсемплинг делается очень просто:
    
`oversample = SMOTE(random_state=RANDOM_STATE)`

`features_train_up, target_train_up = oversample.fit_resample(features_train_enc, target_train)`
    
*предварительно импотрировав `from imblearn.over_sampling import SMOTE`

**Чтобы импортировать imblearn его сначала требуется проинсталлировать. через `!pip install imblearn`
    
</div>


~Теперь с помощью перебора определим сколько раз скопировать положительные ответы, чтобы добиться лучшего результата `F1` от модели:~

In [31]:
#best_result_up = 0
#best_repeats = 0

#for repeats in range(1, 11):

#    features_train_upsampled, target_train_upsampled = upsample(features_train_encoded, target_train, repeats)

#    model = RandomForestClassifier(random_state=12345)

#    model.fit(features_train_upsampled, target_train_upsampled)

#    predicted_valid = model.predict(features_valid_encoded)
    
#    result_up = f1_score(target_valid, predicted_valid)
    
#    if result_up > best_result_up:
#        best_result_up = result_up
#        best_repeats = repeats
        
#print(' F1-мера наилучшей модели случайного леса на валидационной выборке:', best_result_up,
#      '\n При количестве повторов при увеличении выборки:', best_repeats)

С помощью функции SMOTE поправим дисбаланс классов в выборке:

In [32]:
oversample = SMOTE(random_state=12345)
features_train_u, target_train_u = oversample.fit_resample(features_train_encoded, target_train)

Теперь у нас есть сбалансированные выборки для дальнейшего исследования:

<div class="alert alert-block alert-danger">
<h2><s> Комментарий ревьюера <a class="tocSkip"></h2>
  
<b>На доработку🤔:</b>
 
Подбирать баланс не стоит. Аналитически самым удачным будет вариант около баланса (1:1), то есть, у нас это будет коэффициент = 4. Почему же у нас получилось иначе? Потому что мы подбирали баланс классов под необученную модель. Следует сделать баланс 1:1, а затем заново подобрать гиперпараметры. + при подборе баланса классов выше вероятность переобучиться.
    
    
    
В автоматических функциях для апсемплинга типа SMOTE нас даже не спросят под подбор баланса, он по умолчанию там 1:1.

</div>


<div class="alert alert-info">
✍️ Попробовал применить SMOTE вместо ручного апсемплинга (получается и проще и красивее) и убрал перебор баланса на модели

<div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

Здесь аналогично.
       
</div>



<div class="alert alert-block alert-danger">
<h2><s> Комментарий ревьюера <a class="tocSkip"></h2>
  
<b>На доработку🤔:</b>
 
Даунсемплинг мы применяем к изначальным выборкам, а не к апсемплированным. 
    

</div>

<div class="alert alert-info">
✍️ Убрал даунсемплинг
    

<div class="alert alert-block alert-danger">
<h2><s> Комментарий ревьюера <a class="tocSkip"></h2>
  
<b>На доработку🤔:</b>
 
Важно попробовать с апсемплингом все ранее выбранные модели (также Дерево и регрессию).

</div>


<div class="alert alert-info">
✍️ Ниже прогнал после балансировки все модели, что были в части исследования задачи

Снова начнем исследование с модели дерева решений `model_dtc`:

In [33]:
best_result_dtc = 0
best_depth_dtc = 0

for depth in range(1, 11):
    model_dtc = DecisionTreeClassifier(random_state=12345, max_depth=depth)
    model_dtc.fit(features_train_u, target_train_u)
    predictions_valid_dtc = model_dtc.predict(features_valid_encoded)
    result_valid_dtc = f1_score(target_valid, predictions_valid_dtc)
    if result_valid_dtc > best_result_dtc:
        best_result_dtc = result_valid_dtc
        best_depth_dtc = depth
        
    
print(' F1-мера наилучшей модели дерева решений на валидационной выборке:', best_result_dtc, 
'\n При максимальной глубине дерева:', best_depth_dtc)

 F1-мера наилучшей модели дерева решений на валидационной выборке: 0.5956964892412231 
 При максимальной глубине дерева: 8


На сбалансированных выборках модель показывает результат получше.

Теперь проверим модель логистической регрессии `model_lr`:

In [34]:
best_result_lr = 0
best_itr_lr = 0

for itr in range(100, 501, 10):
    model_lr = LogisticRegression(random_state=12345, solver='liblinear', max_iter=itr)
    model_lr.fit(features_train_u, target_train_u)
    predictions_valid_lr = model_lr.predict(features_valid_encoded)
    result_valid_lr = f1_score(target_valid, predictions_valid_lr)
    if result_valid_lr > best_result_lr:
        best_result_lr = result_valid_lr
        best_itr_lr = itr

print(' F1-мера наилучшей модели логистической регрессии на валидационной выборке:', best_result_lr,
      '\n При количестве итераций:', best_itr_lr)

 F1-мера наилучшей модели логистической регрессии на валидационной выборке: 0.5079962370649106 
 При количестве итераций: 100


Модель логистической регресии на сбалансированных выборках показывает гораздо более хороший результат

Теперь на этих сбалансированных выборках обучим модель случайного леса `model_rfc`, показавшую до этого самый лучший результат, чтобы определить какие гиперпараметры задать модели:

In [35]:
best_result_rfc = 0
best_est_rfc = 0
best_depth_rfc = 0

for est in range(30, 41):
    for depth in range(1, 16):
        model_rfc = RandomForestClassifier(random_state=12345, n_estimators=est, max_depth=depth)
        model_rfc.fit(features_train_u, target_train_u)
        predictions_valid_rfc = model_rfc.predict(features_valid_encoded)
        result_valid_rfc = f1_score(target_valid, predictions_valid_rfc)
        if result_valid_rfc > best_result_rfc:
            best_result_rfc = result_valid_rfc
            best_est_rfc = est
            best_depth_rfc = depth
            

        
print(' F1-мера наилучшей модели случайного леса на валидационной выборке:', best_result_rfc,
      '\n При количестве деревьев:', best_est_rfc, 'При глубине деревьев', best_depth_rfc)

 F1-мера наилучшей модели случайного леса на валидационной выборке: 0.655483870967742 
 При количестве деревьев: 36 При глубине деревьев 11


После проведения балансировки модель случайного леса снова показывает себя лучше в предсказаниях.

Сохраним полученную лучшую модель с полученными гиперпараметрами для проверки ее на тестовой выборке.

In [36]:
best_model = RandomForestClassifier(random_state=12345, n_estimators=36, max_depth=11)
best_model.fit(features_train_u, target_train_u)

RandomForestClassifier(max_depth=11, n_estimators=36, random_state=12345)

После корректировки дисбаланса класса с помощью функции SMOTE получилось улучшить показатель `F1` наших моделей. Более хороший показали и модель дерева решений и модель логической регрессии, однако как и при иссследовании до корректировки дисбаланса лучше других себя показала модель случайного леса. Мы переобучили ее и подобрали новые гиперпараметры (количество деревьев 36, глубина 11), теперь пришло время опробовать ее на тестовой выборке.

## Тестирование модели

Протестируем нашу обученную модель на тестовой выборке:

In [37]:
predictions_test = best_model.predict(features_test_encoded)
f1_score(target_test, predictions_test)

0.601078167115903

<div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

Здесь не стоит смешивать апсемплинг и даунсемплинг. Важно выбрать один из способов, применить верный коэффициент и обучить модель с ранее найденными оптимальными гиперпараметрами для способа борьбы с дисбалансом классов.
       
</div>


Модель показала себя хуже, чем на валидационной выборке, но F1-мера превышает необходимый уровень в 0.59.

Для проверки на адекватность создадим dummy-модель, обучим ее на подготовленных учебных выборках и узнаем ее F1-меру для тестовой выборки:

In [38]:
dummy_model = DummyClassifier(strategy='constant', constant=1)
dummy_model.fit(features_train_u, target_train_u)
f1_score(dummy_model.predict(features_test_encoded), target_test)

0.3235023041474654

Проверку на адекватность наша модель прошла.

Теперь изучим модель с помощью допольнительной оценки AUC-ROC:

In [39]:
probabilities_test = best_model.predict_proba(features_test_encoded)
probabilities_one_valid = probabilities_test[:, 1]

print(roc_auc_score(target_test, probabilities_one_valid))

0.855975919327418


Здесь модель тоже проходит проверку, показатель выше критического 0.5.

Так как для нас важно обнаружнить тех клиентов, которые планируют уйти, то проверим нашу модель по метрике полноты:

In [40]:
recall_score(target_test, predictions_test)

0.6353276353276354

**Вывод по проекту:**

Мы проверили работу нескольких моделей и выбрали из них наилучшую модель, это модель случайного леса с гиперпараметрами: количество деревьев 36, глубина 11.

На тестово выборке она показала результат F1-меры равный 0.603, отвечающий нашим требованиям и гораздо более высокий чем у dummy-модели 0.323.

Показатель ROC AUC модели составляет 0,856, что говорит о том, что из предсказанных моделью клиентов готовых рассторгнуть договор 85% действительно собираются это сделать.

Так же проверен на тестовой выборке показатель полноты, он равен 0.635, это означает, что из тех клиентов, которые планируют уходить, модель обнаруживает 63%.

<div class="alert alert-block alert-danger">
<h2><s> Комментарий ревьюера <a class="tocSkip"></h2>
  
<b>На доработку🤔:</b>
 
Необходимо дополнить общий вывод по проекту: описать лучшую модель (гиперпараметры), ее результаты на тестовой выборке, сравнить модель с константной, прокомментировать, как модель решает задачу бизнеса: здесь можно подсчитать recall и объяснить, какую долю клиентов на отток мы детектируем верно.

</div>

<div class="alert alert-info">
✍️ Дописал

<div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

Здесь в дамми-модели я бы предсказал везде «1». Условно говоря, если предсказываем ноль, моделируем ситуацию, когда акцию с бонусами для удержания клиентов решили не проводить. Если предсказываем единицу − решили раздать бонусы вообще всем клиентам. Если в дамми-модели предсказываем по единице, можно подсчитать F1, она не будет зануляться.
    
`dummy_model = DummyClassifier(strategy='constant', constant=1)
dummy_model.fit(features_train, target_train)
f1_score(dummy_model.predict(features_test), target_test)`
    
Импорт: from sklearn.dummy import DummyClassifier

       
</div>


<div class="alert alert-info">
✍️ Выше добавил дамми-модель, спасибо за объяснение с предсказанием единицы

<div style="border:solid Chocolate 2px; padding: 40px">

**Общий вывод по проекту**
    
Спасибо за твой проект! Есть несколько важных моментов, которые необходимо улучшить. Их ты найдешь в разноцветных комментариях.
    
**Отмечу положительные моменты проекта🚀**
    
* Хорошая структура проекта. Выбор моделей.
* Отличная часть с предобработкой данных.
* Есть пояснения к шагам, код с высокой читаемостью. 
* Хороший результат на тестовой выборке.
    
**На что стоит обратить внимание🤔**
    
* Нужно масштабировать данные после разбиения на выборки
* Гиперпараметры моделеям подбираем в каждом случае применения методики борьбы с дисбалансом и в исследовании моделей без учета дисбаланса.
* Апсемплинг не нужно сочетать с даунсемплингом.
* Апсемпилнг нужно довести до баланса классов (1:1, умножаем редкий класс на 4).
* Необходимо везде оставлять промежуточные выводы и дополнить итоговый вывод.
* Эти и остальные замечания я также написал по ходу проекта.
    
**Желаю удачи и жду твой проект на повторное ревью!**😉
    
    
</div>


<div class="alert alert-info">
✍️ Переделал. Спасибо за советы и ревью 🤝


<div style="border:solid Chocolate 2px; padding: 40px">

**Комментарий ко второй проверке**

Спасибо за отличную доработку. Критичных комментариев нет. 

Желаю удачи в дальнейшей учебе!

</div>


## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Выполнен шаг 1: данные подготовлены
- [x]  Выполнен шаг 2: задача исследована
    - [x]  Исследован баланс классов
    - [x]  Изучены модели без учёта дисбаланса
    - [x]  Написаны выводы по результатам исследования
- [x]  Выполнен шаг 3: учтён дисбаланс
    - [x]  Применено несколько способов борьбы с дисбалансом
    - [x]  Написаны выводы по результатам исследования
- [x]  Выполнен шаг 4: проведено тестирование
- [x]  Удалось достичь *F1*-меры не менее 0.59
- [x]  Исследована метрика *AUC-ROC*